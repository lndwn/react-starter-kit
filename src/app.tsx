import React from 'react'
import styled from 'styled-components'
import { Flex } from './components'

const Heading = styled.h1`
  color: ${(props) => props.theme.colors.text[2]};
`

interface IAppProps {
  heading?: string
}

export const App = ({ heading = 'Hello World' }: IAppProps) => {
  return (
    <Flex alignItems="center" justifyContent="center">
      <Heading>{heading}</Heading>
    </Flex>
  )
}
