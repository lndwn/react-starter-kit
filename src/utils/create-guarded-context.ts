import {
  createContext as createReactContext,
  useContext as useReactContext,
} from 'react'

export const createContext = <C extends unknown | null>() => {
  const context = createReactContext<C | undefined>(undefined)
  const useContext = () => {
    const value = useReactContext(context)
    if (value === undefined) {
      throw new Error('Consumer must be inside a Provider with a value')
    }
    return value
  }

  const WithContextValue = (props: {
    children: (provided: C) => JSX.Element
  }) => props.children(useContext())

  return [useContext, context.Provider, WithContextValue] as const
}
