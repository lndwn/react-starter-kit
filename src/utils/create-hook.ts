import { forwardRef } from 'react'

export function createHook<
  Props = unknown | unknown[],
  ProvidedProps = unknown,
  Ref = unknown
>(hook: (props: Props) => ProvidedProps) {
  const WithHook = (
    props: Props & { children: (props: ProvidedProps) => JSX.Element }
  ) => props.children(hook(props))

  const WithRefHook = forwardRef<
    Ref,
    Props & { children: (props: ProvidedProps) => JSX.Element }
  >((props, ref) => props.children(hook({ ref, ...props })))

  return [hook, WithHook, WithRefHook] as const
}
