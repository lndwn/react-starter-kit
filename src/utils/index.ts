export * from './use-document-title'
export * from './use-preferred-color-scheme'
