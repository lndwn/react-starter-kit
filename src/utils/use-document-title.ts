import * as React from 'react'

export const useDocumentTitle = (title?: string) => {
  const [documentTitle, setDocumentTitle] = React.useState<string | undefined>(
    title
  )

  React.useEffect(() => {
    if (title) setDocumentTitle(title)
  }, [title])

  React.useEffect(() => {
    if (documentTitle) document.title = documentTitle
  }, [documentTitle])
}
