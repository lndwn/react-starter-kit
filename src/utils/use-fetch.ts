import * as React from 'react'
import { useInterval } from './use-interval'

interface UseFetchOptions {
  init: RequestInit
  parseAs?: 'json' | 'text' | 'blob' | 'arrayBuffer'
  poll?: boolean
  interval?: number
}

export const useFetch = <ResBody>(url: string, options: UseFetchOptions) => {
  const [isLoading, setIsLoading] = React.useState(true)
  const [response, setResponse] = React.useState<Response | null>(null)
  const [responseBody, setResponseBody] = React.useState<ResBody | null>(null)
  const [responseError, setResponseError] = React.useState<Error | null>(null)
  const [bodyError, setBodyError] = React.useState<Error | null>(null)
  const controller = React.useMemo(() => new AbortController(), [url])
  const request = React.useMemo(
    () => new Request(url, { ...options.init, signal: controller.signal }),
    [url]
  )

  const getResponse = async () => {
    try {
      setIsLoading(true)
      setResponse(await fetch(request))
    } catch (err) {
      setIsLoading(false)
      setResponseError(err)
    }
  }

  React.useEffect(() => {
    getResponse()
  }, [url])

  useInterval(() => getResponse(), {
    interval: options?.interval ?? 5000,
    isEnabled: options?.poll ?? false,
  })

  const getBody = async () => {
    if (response) {
      try {
        setResponseBody(await response[options?.parseAs ?? 'json']())
        setIsLoading(false)
      } catch (err) {
        setIsLoading(false)
        setBodyError(err)
      }
    }
  }

  React.useEffect(() => {
    getBody()
  }, [response])

  return {
    isLoading,
    response,
    responseBody,
    responseError,
    bodyError,
    controller,
  }
}
