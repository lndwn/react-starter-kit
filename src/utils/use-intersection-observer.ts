import * as React from 'react'

interface UseIntersectionObserverProps {
  options: IntersectionObserverInit
  callback: IntersectionObserverCallback
}

export const useIntersectionObserver = (
  props: UseIntersectionObserverProps
) => {
  const ref = React.useRef<HTMLElement | null>(null)
  React.useEffect(() => {
    const observer = new IntersectionObserver(props.callback, props.options)
    if (ref.current) {
      observer.observe(ref.current)
    }
    return () => observer.disconnect()
  }, [props])

  return ref
}

interface WithIntersectionObserverProps extends UseIntersectionObserverProps {
  children: (
    ref?: React.MutableRefObject<HTMLElement | null>
  ) => React.ReactNode
}

export const WithIntersectionObserver = (
  props: WithIntersectionObserverProps
) => {
  const ref = useIntersectionObserver({
    options: props.options,
    callback: props.callback,
  })

  return props.children(ref)
}
