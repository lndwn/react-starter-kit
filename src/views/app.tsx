import { GlobalStyle } from 'style/global-style'
import { ThemeProvider } from 'styled-components'
import { defaultTheme } from 'theme'

export const App = () => {
  return (
    <ThemeProvider theme={defaultTheme}>
      Hello World
      <GlobalStyle />
    </ThemeProvider>
  )
}
