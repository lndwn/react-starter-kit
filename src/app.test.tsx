import React from 'react'
import { render } from '@testing-library/react'
import { App } from './app'
import { TestFrame } from './utils/test-frame'

describe('index.tsx', () => {
  it('renders with theme context', () => {
    const { container, getByText } = render(
      <TestFrame>
        <App />
      </TestFrame>
    )

    expect(getByText(/hello world/i))
    expect(container.firstChild).toMatchSnapshot()
  })
})
