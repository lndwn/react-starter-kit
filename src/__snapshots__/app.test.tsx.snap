// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`index.tsx renders with theme context 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
}

.c1 {
  color: #808080;
}

@media (prefers-reduced-motion:reduce) {

}

<div
  class="boxsc__Box-sc-1g017hi-0 c0"
>
  <h1
    class="c1"
  >
    Hello World
  </h1>
</div>
`;
